[clang] Add clang::preferred_type attribute for bitfields

Endilll opened the pull request:
This attribute allows user to specify type of the bitfield that will be emitted to debug info without affecting semantics of the program. Since it doesn't affect semantics, this attribute can be safely ignored by other compilers.

This is useful when user is forced to use the same type for all bitfields in a class to get better [layout](https://godbolt.org/z/ovWqzqv9x) and [codegen](https://godbolt.org/z/bdoqvz9e6) from MSVC, because it allows debuggers to interpret the value of bitfield in the most human-friendly way (e.g. when value actually comes from an enum). This is driven by my work on LLDB formatters for Clang. I have two use cases for this:
```cpp
namespace Clang {
class Type {
  enum TypeClass { ... };
  struct TypeBitfields {
    [[clang::debug_info_type(clang::Type::TypeClass)]] unsigned TC: 8;
    [[clang::debug_info_type(bool)]] mutable unsigned FromAST : 1;
  };
};
}
```

Endilll committed:
[clang] Add clang::debug_info_type attribute

Endilll committed:
Add newline to the test

Endilll commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,28 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
+    QualType BitfieldType = llvm::cast<FieldDecl>(D)->getType();
+    QualType EnumUnderlyingType = type->getAs<EnumType>()->getDecl()->getIntegerType();
+    if (EnumUnderlyingType != BitfieldType) {
+      S.Diag(AL.getLoc(), diag::warn_attribute_underlying_type_mismatch) << EnumUnderlyingType << type << BitfieldType;
```
I intend to pass a first argument location instead of `AL.getLoc()`, but `AL.getArg(0)` triggers assertion that the index is out of bounds. What I am missing here?

Endilll replied:
@AaronBallman @erichkeane Can I request some attention here?

AaronBallman replied:
Sorry, I completely missed this comment because it doesn't show up in the Files tab (how helpful).

`getArg()` is used to get arguments that are either an `Expr *` or an `IdentifierInfo *`, but this attribute accepts a type argument which is neither of those things. We have `getTypeArg()` (which is a pretty bad code smell because this only allows *one* type argument for an attribute -- it would probably be a reasonable idea to refactor this so it works through `getArg()` at some point. CC @erichkeane). This gets you a `ParsedType` which you'll need to call `Sema::GetTypeFromParser()` on to get the actual type information for, and that includes a `TypeSourceInfo *` which will have location information for the type. In the current code, you can use `ParmTSI` to get the `TypeLoc` for this.

Endilll replied:
> In the current code, you can use ParmTSI to get the TypeLoc for this.

It's working. Thank you!

Endilll committed:
Run clang-format

dwblaikie commented:
> This is useful when user is forced to use the same type for all bitfields in a class to get better [layout](https://godbolt.org/z/ovWqzqv9x) and [codegen](https://godbolt.org/z/bdoqvz9e6) from MSVC

Does this issue not apply to other platforms? (I didn't think you could bitfield enums)

I'd think maybe rather than adding an attribute to workaround MSVC's missing support - code could #ifdef in a way that's ABI compatible, maybe? (though I admit that's a subtle thing to get right & the cost for getting it wrong is significant)

Endilll commented:
> Does this issue not apply to other platforms?

As far as I understand, MS ABI forces compilers to insert padding between bitfields more often than Itanium. Probably because of that GCC and Clang are able to merge separate loads of bitfields into single load on Linux.

> I'd think maybe rather than adding an attribute to workaround MSVC's missing support - code could #ifdef in a way that's ABI compatible, maybe? (though I admit that's a subtle thing to get right & the cost for getting it wrong is significant)

It's technically possible to `#ifdef` this for Linux, but it's not going to work on Windows, because we have to show the compiler that bitfields have the same type (for Clang it's usually `unsigned`). Otherwise we get bad layouts and bad codegen, which is an unacceptable regression.

Even if we focus on Linux, I'd rather pay with lots of boilerplate in LLDB formatter (lower estimate is 4 lines per every single bitfield), than litter Clang headers with `#ifdef`s for the sake of debuggers.

pogo59 commented:
> Does this issue not apply to other platforms?

There is a deliberate layout/ABI choice that MSVC made ages ago and will never change. It will only pack bitfields into the same word if the neighboring bitfield declarations have the same base type. This is relatively well known.

That said, being able to attach an enum (or whatever) to the bitfield does seem handy, for any target.

dwblaikie commented:
> Wouldn't it be better to go the other way around? i.e. have a `[[clang::compressed_bitfield]]` (or whatever) which influences the ABI, so it's possible to do stuff like
> 
> ```c++
> [[clang::compressed_bitfield]] bool IsSomething : 1;
> [[clang::compressed_bitfield]] MyEnum Whatever : 3;
> [[clang::compressed_bitfield]] int MoreStuff : 4;
> ```
> 
> which the current approach doesn't allow.

The issue is that MSVC wouldn't know/recognize/implement this attribute - so clang couldn't do that without breaking ABI compatibility with MSVC.

AaronBallman commented:
> > > Wouldn't it be better to go the other way around? i.e. have a `[[clang::compressed_bitfield]]` (or whatever) which influences the ABI, so it's possible to do stuff like
> > > ```
> > > [[clang::compressed_bitfield]] bool IsSomething : 1;
> > > [[clang::compressed_bitfield]] MyEnum Whatever : 3;
> > > [[clang::compressed_bitfield]] int MoreStuff : 4;
> > > ```
> > > 
> > > 
> > > which the current approach doesn't allow.
> > 
> > 
> > The issue is that MSVC wouldn't know/recognize/implement this attribute - so clang couldn't do that without breaking ABI compatibility with MSVC.
> 
> That's not important for everybody, and you could still `#if` around that code like
> 
> ```c++
> #if __has_cpp_attribute(clang::compressed_bitfield)
> #define BITFILED_T(ActualType, ABIType) static_assert(check_abi_compat<ActualType, ABIType>()); [[clang::compressed_bitfield]] ActualType
> #else
> #define BITFIELD_T(ActualType, ABIType) ABIType
> ...
> ```
> 
> Sure, it's not super pretty, but it gets you there and allows you to do more stuff. We could also ask the MSVC folks whether there would be interest to add such an attribute, since there is clearly interest in having a good ABI and stronger typing (CC @StephanTLavavej @CaseyCarter).

I think that's a somewhat orthogonal feature to the one @Endilll is proposing. While it might make sense to ask the MSVC and GCC folks if they'd be willing to entertain a similar attribute, the implementation landscape is much wider in C and those implementations may have similar ABI issues. If an implementation ignores the `compressed_bitfield` attribute, you get potential ABI issues; if an implementation ignores the `debug_info_type` attribute, you get the default debugging experience but no change to codegen behavior. So I think there may be room for both features.

We have a `preferred_name` attribute already for specifying "please display this name instead of that name" in diagnostics; this seems somewhat similar in nature in that it's "please display this type instead of that bit-field type" in the debugger, so mayyybe we want to name this `preferred_type` instead? If this could also be useful for diagnostics which mention types and can be triggered with a bit-field, that might also be a reason to use a different name (I'm not certain we have many/any such diagnostics today though.) I'm not opposed to the current name, but more thinking that it might be too specific and we may want a path to more generalization in the future.

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,30 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
+  }
+
+  TypeSourceInfo *ParmTSI = nullptr;
+  QualType type = S.GetTypeFromParser(AL.getTypeArg(), &ParmTSI);
```
```suggestion
  QualType QT = S.GetTypeFromParser(AL.getTypeArg(), &ParmTSI);
```
(or any other name that meets the coding style, except `Type` because.. that's a type.)

Endilll replied:
Fixed

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,30 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
+  assert(ParmTSI && "no type source info for attribute argument");
+
+  if (type->isEnumeralType()) {
+    QualType BitfieldType = llvm::cast<FieldDecl>(D)->getType();
```
```suggestion
    QualType BitfieldType = cast<FieldDecl>(D)->getType();
```

Endilll replied:
Fixed

erichkeane commented:
> > > > Wouldn't it be better to go the other way around? i.e. have a `[[clang::compressed_bitfield]]` (or whatever) which influences the ABI, so it's possible to do stuff like
> > > > ```
> > > > [[clang::compressed_bitfield]] bool IsSomething : 1;
> > > > [[clang::compressed_bitfield]] MyEnum Whatever : 3;
> > > > [[clang::compressed_bitfield]] int MoreStuff : 4;
> > > > ```
> > > > 
> > > > 
> > > > which the current approach doesn't allow.
> > > 
> > > 
> > > The issue is that MSVC wouldn't know/recognize/implement this attribute - so clang couldn't do that without breaking ABI compatibility with MSVC.
> > 
> > 
> > That's not important for everybody, and you could still `#if` around that code like
> > ```
> > #if __has_cpp_attribute(clang::compressed_bitfield)
> > #define BITFILED_T(ActualType, ABIType) static_assert(check_abi_compat<ActualType, ABIType>()); [[clang::compressed_bitfield]] ActualType
> > #else
> > #define BITFIELD_T(ActualType, ABIType) ABIType
> > ...
> > ```
> > 
> > 
> > Sure, it's not super pretty, but it gets you there and allows you to do more stuff. We could also ask the MSVC folks whether there would be interest to add such an attribute, since there is clearly interest in having a good ABI and stronger typing (CC @StephanTLavavej @CaseyCarter).
> 
> I think that's a somewhat orthogonal feature to the one @Endilll is proposing. While it might make sense to ask the MSVC and GCC folks if they'd be willing to entertain a similar attribute, the implementation landscape is much wider in C and those implementations may have similar ABI issues. If an implementation ignores the `compressed_bitfield` attribute, you get potential ABI issues; if an implementation ignores the `debug_info_type` attribute, you get the default debugging experience but no change to codegen behavior. So I think there may be room for both features.
> 
> We have a `preferred_name` attribute already for specifying "please display this name instead of that name" in diagnostics; this seems somewhat similar in nature in that it's "please display this type instead of that bit-field type" in the debugger, so mayyybe we want to name this `preferred_type` instead? If this could also be useful for diagnostics which mention types and can be triggered with a bit-field, that might also be a reason to use a different name (I'm not certain we have many/any such diagnostics today though.) I'm not opposed to the current name, but more thinking that it might be too specific and we may want a path to more generalization in the future.

While I agree that a 'pack-at-the-bit-level' type attribute might be valuable, I too see that as a 'different' problem.  There is plenty of code (even in Clang itself) where and Enum is being stored as a bitfield, and more interesting debugging information could be generated for it.  I don't mind the name, I think there IS a nice level of symmetry to `preferred_type`, which I think ALSO has the advantage of being useful (by name!) for non-debug diagnostics.

One thing I REALLY hate about our pattern of using bitfields for enums, is that it becomes pretty trivial to add an entry to an enum and forget to increase the size of the bitfield.  I think such a attribute could cause a diagnostic of `bitfield says it holds an ENUM type, but can't store all the values`, which would be INCREDIBLY valuable (not that I'm pushing that this patch IMPLEMENT that, but more that I'm saying such a name COULD do that, and probably should as well).

So in conclusion, I don't mind the name, but also prefer something more general.

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -107,6 +107,10 @@ def NonBitField : SubsetSubject<Field,
 
+def BitField : SubsetSubject<Field,
+                             [{S->isBitField()}],
+                             "bit-field non-static data members">;
```
I don't know how much this text helps here, the diagnostic specifying 'non static' seems a little silly.  The `NonBitField` one is important because it applies to non-static data members that ARE NOT bitfields.  But in this case, 'must be a bitfield' also implies 'non static'.

Endilll replied:
Fixed

AaronBallman commented on clang/test/CodeGen/debug-info-debug-info-type.cpp:
```diff
@@ -0,0 +1,14 @@
+// RUN: %clang -target x86_64-linux -g -S -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -verify -DMISMATCH %s
```
I think we should split this into two test: one for testing what debug info is generated (here in CodeGen) and another one for testing diagnostic behavior in Sema. The CodeGen test should be clean of diagnostics.

Another test I think we should add would be testing qualifier mismatches (where the bit-field is `const`-qualified to make sure that doesn't show up as a mismatch.

Endilll replied:
Tests are split now. `const` case is now checked and working as expected (qualifiers are ignored).

philnik777 commented:
> One thing I REALLY hate about our pattern of using bitfields for enums, is that it becomes pretty trivial to add an entry to an enum and forget to increase the size of the bitfield. I think such a attribute could cause a diagnostic of `bitfield says it holds an ENUM type, but can't store all the values`, which would be INCREDIBLY valuable (not that I'm pushing that this patch IMPLEMENT that, but more that I'm saying such a name COULD do that, and probably should as well).
> 
> So in conclusion, I don't mind the name, but also prefer something more general.

Couldn't we add a warning for that in general? Seems like a pretty useful one to me.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7219,6 +7219,18 @@ its underlying representation to be a WebAssembly ``funcref``.
+facilitate compiler optimizations. But this also causes underlying type to be
+emitted in debug information, making it hard for debuggers to map bitfield's
+value back to enumeration. This attribute helps with this.
+  }];
```
This whole bit of documentation reads REALLY awkwardly and has a couple of sentences that are run-on or not-full sentences.  AS a strawman:

"
Attribute `[[<whateverwenameit>]]` provides a mechanism to alter the type in debug information of a bitfield.  This is useful as it enables a better debug experience in cases where, for layout or other performance purposes, an enumeration value is best represented in a structure as a bitfield. By changing the type in the debug information, the layout is untouched, but debuggers are able to map the bitfield's values back to the enumeration values they represent.
"

Endilll replied:
Applied the text suggested by Aaron

erichkeane commented:
> > One thing I REALLY hate about our pattern of using bitfields for enums, is that it becomes pretty trivial to add an entry to an enum and forget to increase the size of the bitfield. I think such a attribute could cause a diagnostic of `bitfield says it holds an ENUM type, but can't store all the values`, which would be INCREDIBLY valuable (not that I'm pushing that this patch IMPLEMENT that, but more that I'm saying such a name COULD do that, and probably should as well).
> > So in conclusion, I don't mind the name, but also prefer something more general.
> 
> Couldn't we add a warning for that in general? Seems like a pretty useful one to me.

Not without SOME sort of psychic-level analysis from Sema.  Without some level of mapping of "I mean for this to hold the values of this Enum", we don't have the information to do so.  We cannot take advantage of assignments (though a similar warning DOES exist in the 'range' checking) since it might be intended that the field does some level of 'conversion' between enums.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,30 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
+    QualType BitfieldType = llvm::cast<FieldDecl>(D)->getType();
+    QualType EnumUnderlyingType =
+        type->getAs<EnumType>()->getDecl()->getIntegerType();
+    if (EnumUnderlyingType != BitfieldType) {
```
> A few things -- I think we should look at the canonical type so that we don't run into issues with typedefs. e.g.,
> 
> ```
> enum E { Zero, One };
> typedef int Foo;
> struct S {
>   [[clang::debug_info_type(E) Foo field : 1;
> };
> ```
> 
> where the enum's underlying type is `int` but the bit-field's type is a typedef.
> 
> Another case we should consider would be whether we want to allow signed unsigned mismatches, as in:
> 
> ```
> enum E { Invalid = -1, ValidValue, OtherValidValue, YetAnotherValidValue };
> struct S {
>   [[clang::debug_info_type(E) unsigned field : 3;
> };
> ```
> 
> and only prevent situations where the specified type is not an enumeration or integral type.

This ALSO makes me wonder if forcing it to be an enum type is necessary.  I can see potential value of a "OpaqueValueOfStruct" storage type thing, that perhaps we should just 'trust' that the user is doing something sensible with it.

  I think in the signed/unsigned mismatch, we should allow it (thanks to the layout rules).  That is, consider:

```
enum SIGNED {Invalid = -1, Valid };
enum UNSIGNED {Valid, INVALID = 0xF };

struct S {
  [[attr_name(SIGNED)]]
  unsigned SignedThing : 4;
  [[attr_name(UNSIGNED)]]
  unsigned UnsignedThing : 4;
};
```
They HAVE to be the same type else they don't get packed into the same byte, but you have 1 enum that can represent a negative value.

Endilll replied:
> and only prevent situations where the specified type is not an enumeration or integral type.

> This ALSO makes me wonder if forcing it to be an enum type is necessary.

I intentionally haven't been restricting the set of types that can be passed as the argument to `preferred_type`. This makes even more sense now that we are going with more generic name. I'm just making sure that enum-related diagnostics are triggered only when enum was passed by the user.

> Another case we should consider would be whether we want to allow signed unsigned mismatches:

> I think in the signed/unsigned mismatch, we should allow it (thanks to the layout rules)

> 1 enum that can represent a negative value

I disagree. We shouldn't pretend that our enums can hold negative values, while we store them in unsigned bit-fields (save for rare exceptions, if any). Now that we are not bound by C++98 compilers, I believe we should propagate underlying type of our enums from bit-field type they are stored in (usually `unsigned`, `uint64_t` for `DeclBase.h`), and deal with negative values at enumerator declaration (I guess we'll have to explicitly cast them to the underlying type). I can do that while adding `clang::preferred_type` to bit-fields.

Ideally I'd propagate bit-field width as well, but that doesn't seem possible at the moment: https://github.com/llvm/llvm-project/pull/69104#issuecomment-1770189229

AaronBallman replied:
I misread an earlier comment of yours (I thought you were saying that the underlying type of the enum and the bit-field needed to match, not that we're pulling information from one and applying it to the other.)

AaronBallman replied:
I think we should ignore signed vs unsigned mismatch on the enum vs bit-field types because I'm concerned about how chatty the diagnostic will be with portable code. MSVC and Clang/GCC make different decisions about whether an unscoped enumeration has a signed or unsigned underlying type, so it makes me wary.

We can always revisit this decision in the future if we find a better heuristic or more worrisome code in the wild.

WDYT?

Endilll replied:
After chatting with Aaron offline, I used his solution to ignore sign-ness when comparing bit-field type against enum enumeration type.

Endilll replied:
With `typedef` and sign-ness issues fixed, I consider this done.

philnik777 commented:
> > > One thing I REALLY hate about our pattern of using bitfields for enums, is that it becomes pretty trivial to add an entry to an enum and forget to increase the size of the bitfield. I think such a attribute could cause a diagnostic of `bitfield says it holds an ENUM type, but can't store all the values`, which would be INCREDIBLY valuable (not that I'm pushing that this patch IMPLEMENT that, but more that I'm saying such a name COULD do that, and probably should as well).
> > > So in conclusion, I don't mind the name, but also prefer something more general.
> > 
> > 
> > Couldn't we add a warning for that in general? Seems like a pretty useful one to me.
> 
> Not without SOME sort of psychic-level analysis from Sema. Without some level of mapping of "I mean for this to hold the values of this Enum", we don't have the information to do so. We cannot take advantage of assignments (though a similar warning DOES exist in the 'range' checking) since it might be intended that the field does some level of 'conversion' between enums.

I meant warning on something like https://godbolt.org/z/Ma17xjjc5. That doesn't seem like it should be that hard.

Endilll commented:
@erichkeane I guess you'd be pretty happy if our enums were declared the following way:
```cpp
enum E : unsigned _BitInt(4) {
  e0 = 16,
};
```
Then you'd be warned on the spot if you have more values than bits can hold, and bit-fields side check would be trivial to implement:
```
<source>:3:8: error: enumerator value evaluates to 16, which cannot be narrowed to type 'unsigned _BitInt(4)' [-Wc++11-narrowing]
  e0 = 16,
```
Unfortunately, Clang 14+ is the only major implementation that accepts this: https://godbolt.org/z/9rn87d8hc

Endilll committed:
Address feedback from reviewers

- Rename the attribute to `preferred_type`
- Add a new diagnostic that checks that `preferred_type(bool)` is attached to bit-field of width 1
- Split CodeGen and Sema tests, add more tests
- Apply all suggestions

Endilll commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7219,6 +7219,18 @@ its underlying representation to be a WebAssembly ``funcref``.
+facilitate compiler optimizations. But this also causes underlying type to be
+emitted in debug information, making it hard for debuggers to map bitfield's
+value back to enumeration. This attribute helps with this.
+  }];
```
Thank you very much for rewriting it!

Endilll commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,30 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
 }
 
+static void handleDebugInfoTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  if (!AL.hasParsedType()) {
+    S.Diag(AL.getLoc(), diag::err_attribute_wrong_number_arguments) << AL << 1;
+    return;
+  }
```
I don't know what do we do in generated boilerplate, but without this check, test that pass no arguments crashes the compiler in this function.

Endilll committed:
Run clang-format

Endilll commented on clang/test/Sema/attr-preferred-type.cpp:
```diff
@@ -0,0 +1,20 @@
+  // expected-error@-1 {{'preferred_type' attribute takes one argument}}
+  [[clang::preferred_type]] unsigned b7 : 2;
+  // expected-error@-1 {{'preferred_type' attribute takes one argument}}
+  [[clang::preferred_type(E, int)]] unsigned b8 : 2;
```
@AaronBallman On the topic of automatically-generated diagnostics:
```
/home/user/endill/llvm-project/clang/test/Sema/attr-preferred-type.cpp:16:28: error: expected ')'
   16 |   [[clang::preferred_type(E, int)]] unsigned b8 : 2;
      |                            ^
      |                            )
/home/user/endill/llvm-project/clang/test/Sema/attr-preferred-type.cpp:16:33: error: expected ','
   16 |   [[clang::preferred_type(E, int)]] unsigned b8 : 2;
      |                                 ^
      |                                 ,
/home/user/endill/llvm-project/clang/test/Sema/attr-preferred-type.cpp:16:30: warning: unknown attribute 'int' ignored [-Wunknown-attributes]
   16 |   [[clang::preferred_type(E, int)]] unsigned b8 : 2;
      |                              ^~~
```
3 diagnostics are issued, 0 says that wrong number of attribute arguments is passed.

AaronBallman replied:
Huh, something got confused there.... it's parsing `int` as though it were another attribute and not an argument to the `preferred_type` attribute. That surprises me!

erichkeane commented:
>I meant warning on something like https://godbolt.org/z/Ma17xjjc5. That doesn't seem like it should be that hard.

Yeah, that one wouldn't be too hard (and isn't a pattern I've ever seen TBH), but doesn't solve the problem at hand, which is that it is very common to need to use the same type for multiple enum values in bitfields, so folks use unsigned, thus the type name is lost.  

>@erichkeane I guess you'd be pretty happy if our enums were declared the following way:

Not really?  That doesn't gain us the 'we must change where this is stored' situation like a `preferred_type` attribute would/could.  We MIGHT be able to get away with a comment on top of that with a list of places that needs to ALSO be changed, but we end up missing similar comments today anyway.

Endilll commented:
>>    @erichkeane I guess you'd be pretty happy if our enums were declared the following way:

>Not really? That doesn't gain us the 'we must change where this is stored' situation like a preferred_type attribute would/could. We MIGHT be able to get away with a comment on top of that with a list of places that needs to ALSO be changed, but we end up missing similar comments today anyway.

I meant `unsigned _BitInt(N)` underlying type to complement `preferred_type`'s check that a bit-field is wide enough. Otherwise such a check will have to incorporate heuristics to prevent false-positives for cases like the following: https://github.com/llvm/llvm-project/blob/180eae1f1e5a08595ed2278d93f01fb321284649/clang/include/clang/AST/DeclarationName.h#L174-L185
One might argue that this particular case might be resolved by transition to `PointerIntPair`, which is true. It's just the only case I was able to find. I have a recollection of an enum that ends with `Size` enumerator, possibly TableGen'ed, but I can't find it.

erichkeane commented:
> > > @erichkeane I guess you'd be pretty happy if our enums were declared the following way:
> 
> > Not really? That doesn't gain us the 'we must change where this is stored' situation like a preferred_type attribute would/could. We MIGHT be able to get away with a comment on top of that with a list of places that needs to ALSO be changed, but we end up missing similar comments today anyway.
> 
> I meant `unsigned _BitInt(N)` underlying type to complement `preferred_type`'s check that a bit-field is wide enough. Otherwise such a check will have to incorporate heuristics to prevent false-positives for cases like the following:
> 
> https://github.com/llvm/llvm-project/blob/180eae1f1e5a08595ed2278d93f01fb321284649/clang/include/clang/AST/DeclarationName.h#L174-L185
> 
> One might argue that this particular case might be resolved by transition to `PointerIntPair`, which is true. It's just the only case I was able to find. I have a recollection of an enum that ends with `Size` enumerator, possibly TableGen'ed, but I can't find it.

We actually DO a similar kind of diagnostic in one of our assignments already I think, which figures out that a value we're storing into doesn't have the number of bits required for the 'highest' and 'lowest' value of an Enum (though I'm having trouble finding it ATM).  This obviously doesn't 'work' in the case of bitmasks, but to the point that it is 'quiet' instead of false-positive.  I'm just suggesting that we can extend this warning to the declarations of bitfield that use this to specify an enumeration type.

Endilll commented:
> See how on the assignment to a bitfield we check to make sure the largest value of the enum will fit in the bitfield? I'm saying I want us to do this EARLIER, on declaration with this attribute.

I totally do. Thank you for pointing out to `-Wbitfield-enum-conversion`! But you should be able to see in my example above that we need a way to exclude an enumerator from this check (in my example it's `[[clang::non_storable]]` applied to `UncommonNameKindOffset`), otherwise we'll get false-positives.

erichkeane commented:
> > See how on the assignment to a bitfield we check to make sure the largest value of the enum will fit in the bitfield? I'm saying I want us to do this EARLIER, on declaration with this attribute.
> 
> I totally do. Thank you for pointing out to `-Wbitfield-enum-conversion`! But you should be able to see in my example above that we need a way to exclude an enumerator from this check (in my example it's `[[clang::non_storable]]` applied to `UncommonNameKindOffset`), otherwise we'll get false-positives.

I don't see why that is necessary?  We don't have anything like that today for `-Wbitfield-enum-conversion`?  I have a hard time seeing the value of it, even for  `-Wbitfield-enum-conversion`, when does someone have an enum value that they don't intend to ever be in the enum?

erichkeane commented:
>In my previous comments (https://github.com/llvm/llvm-project/pull/69104#issuecomment-1771167758, https://github.com/llvm/llvm-project/pull/69104#issuecomment-1771204043) I provided an example of StoredNameKind enum that has enumerators 0 through 8 (9 total, 4 bits to store), but we store it in a bit-fields of width 3, because the last enumerator is there to count other enumerators, and is used for offset purposes. If we apply -Wbitfield-enum-conversion to it as-is, we get a false-positive: https://godbolt.org/z/4xPs6qPTY.

I thought that was a contrived example, but I see where it is from now (note your example is the related but not-exactly-the-same `Wbitfield-constant-conversion`), which brings up an additional concern/thing to deal with for the `non_storable` attribute (that is, do we apply it to assignments? how about exact-assignments like you've done there?).

That is an odd/awkward pattern even so... I think there is still value to the warning as proposed, but I'm now not opposed to using `non_storable` as an escape hatch.

Endilll commented:
> which brings up an additional concern/thing to deal with for the non_storable attribute (that is, do we apply it to assignments? how about exact-assignments like you've done there?).

Conservative approach would be to issue diagnostics based on `non_storable` only when LHS of assignment was checked taking it into account. In other words, LHS is a bit-field with `preferred_type`, so `non_storable` enumerator was excluded when computing how many bits are required to store the enum.

erichkeane commented:
> > which brings up an additional concern/thing to deal with for the non_storable attribute (that is, do we apply it to assignments? how about exact-assignments like you've done there?).
> 
> Conservative approach would be to issue diagnostics based on `non_storable` only when LHS of assignment was checked taking it into account. In other words, LHS is a bit-field with `preferred_type`, so `non_storable` enumerator was excluded when computing how many bits are required to store the enum.

In that case, yes, we'd exclude it.  It is the storing issue we have today that gets funny.  So:

`my_bitfield = VarOfTypeEnumThatWontFitExceptForNonStorable // Don't diagnose, as we don't consider it`
`my_bitfield = EnumThatWontFitExceptForNonStorable::NonStorableItem; // Do we diagnose this?`

I'd lean towards "No" in the 2nd case, the person is clear in what they are doing there.  They should probably get the existing `Wbitfield-constant-conversion` still though.

AaronBallman commented:
> Taking the example above, I think it would have to look the following way to fully complement a check for bit-field width in `preferred_type`:
> 
> ```c++
>  enum StoredNameKind : unsigned _BitInt(3) { 
> ... snip ...
>```
> I think we can robustly diagnose new enumerators if the code is written this way. 

There's some danger here. `_BitInt` is a C23 feature as are enumerations with a fixed underlying type. Enumerations with a fixed underlying type *explicitly* disallow using a bit-precise integer type as the underlying type. See C23 6.7.2.2p4, which says in part, "For all the integer constant expressions which make up the values of the enumeration constants, there shall be a type capable of representing all the values that is a standard or extended signed or unsigned integer type, or char.".

So relying on `_BitInt` as the underlying type isn't going to work in C unless we wish to ignore that constraint. Further, we might wish to extend that constraint to C++ use of `_BitInt` (to ease accidental header compatibility issues) for similar reasons to why it was restricted in C.

I'm still thinking my way through a `non_storable` attribute, but on its face, it seems like it could be overkill. I suspect (but haven't measured!) that there is *way* more code out there that maps enumerations to bit-fields that expect all members of the enumeration to be possible to store than there is code that has a sentinel value not expected to be used. I kind of wonder if the correct answer to those more exceptional situations is "don't use this attribute, you're not doing a straight mapping in this case." Would it perhaps make sense to handle the common path first, see what usage experience finds in the wild, and then consider `non_storable` in the future?

Endilll commented:
> There's some danger here. _BitInt is a C23 feature as are enumerations with a fixed underlying type. Enumerations with a fixed underlying type explicitly disallow using a bit-precise integer type as the underlying type. See C23 6.7.2.2p4, which says in part, "For all the integer constant expressions which make up the values of the enumeration constants, there shall be a type capable of representing all the values that is a standard or extended signed or unsigned integer type, or char.".

Thank you for pointing out to this! But it's not clear to me whether this was an explicit decision to not add bit-precise integers here, or just nobody cared enough to push for that. Because use case we have on our hands seems legit to me.

> Would it perhaps make sense to handle the common path first, see what usage experience finds in the wild, and then consider non_storable in the future?

Yeah, we should collect more evidence once we start sprinkle `preferred_type` around, with relevant diagnostics enabled (and implemented).

AaronBallman commented:
> > There's some danger here. _BitInt is a C23 feature as are enumerations with a fixed underlying type. Enumerations with a fixed underlying type explicitly disallow using a bit-precise integer type as the underlying type. See C23 6.7.2.2p4, which says in part, "For all the integer constant expressions which make up the values of the enumeration constants, there shall be a type capable of representing all the values that is a standard or extended signed or unsigned integer type, or char.".
> 
> Thank you for pointing out to this! But it's not clear to me whether this was an explicit decision to not add bit-precise integers here, or just nobody cared enough to push for that. Because use case we have on our hands seems legit to me.

Explicit decision: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3030.htm#design-bit.precise.integers -- we should add that diagnostic at some point. :-D

The biggest concern I had is that bit-precise integer types do not undergo integer promotions, but enumerations do. So it's unclear what the expectations are in this case. Should `enum Foo : unsigned _BitInt(8);` promote to `int` like `unsigned char` would, or should it not promote like `unsigned _BitInt(8)` would?

Endilll commented:
I filed an issue for this, automatically mentioned above.

Endilll committed:
Fix location for underlying type mismatch diagnostic

erichkeane commented:
> I filed an issue for this, automatically mentioned above.

Fixed!

Endilll committed:
Ignore sign-ness when comparing bitfield type against enum underlying type

Endilll commented:
I'm ignoring signed/unsigned mismatch as @erichkeane and @AaronBallman suggested. The only outstanding aspect is the following diagnostic I added today and haven't received feedback on:
```cpp
  [[clang::preferred_type(bool)]] unsigned b4 : 1;
  [[clang::preferred_type(bool)]] unsigned b5 : 2;
  // expected-warning@-1 {{bit-field that holds a boolean value should have width of 1 instead of 2}}
```

Endilll committed:
Add missing newline

erichkeane commented:
> I'm ignoring signed/unsigned mismatch as @erichkeane and @AaronBallman suggested. The only outstanding aspect is the following diagnostic I added today and haven't received feedback on:
> 
> ```c++
>   [[clang::preferred_type(bool)]] unsigned b4 : 1;
>   [[clang::preferred_type(bool)]] unsigned b5 : 2;
>   // expected-warning@-1 {{bit-field that holds a boolean value should have width of 1 instead of 2}}
> ```

Yeah, I haven't had a good chance to do a huge deep dive on this today, and probably won't until later this week.  While I think that warning is accurate, I somewhat question the value of the 'bool' as working on this type (as, I'm not sure what it really means to put a non-enum here? What types ARE we allowing in this now?).  

I think the wording of the diagnostic is perhaps awkward, so we should bikeshed that as well, though I'm on the fence about the diagnostic existing at all.  On one hand, this would be such a special case, as typically 'too large' for the type isn't an issue for any reason.  On the other, the 'bool' case is perhaps uniquely an issue.

What about if the person prefers the type be 'char', but has it be a size of 9?  THIS is a case where it is reasonable (consider cases where you need to store the result of one of those calls to the C library that returns 'an int that is either a char or -1').

So I guess I'm really questioning how accurate/error-free we could make that diagnostic.  I'm leaning towards "too big of a bitfield should be ok".

Endilll commented:
> While I think that warning is accurate, I somewhat question the value of the 'bool' as working on this type

I'm not sure what you mean by "working" here, but I'd like to highlight that we have hundreds of single-bit bit-fields across Clang that would benefit from `[[clang::preferred_type(bool)]]`.

> as, I'm not sure what it really means to put a non-enum here?

When it was `clang::debug_info_type`, the meaning was as simple as "I want the value of this bit-field to be interpreted as T". Now that we moved to `preferred_type`, there might be more interpretations.

> What types ARE we allowing in this now?

As I mentioned in https://github.com/llvm/llvm-project/pull/69104#discussion_r1365269451, I'm not putting any restrictions on type parameter of the attribute, which makes even more sense for more generic `preferred_type`.

But I'm confused by the fact you are raising this question. In https://github.com/llvm/llvm-project/pull/69104#discussion_r1364432624, you wrote:
> I can see potential value of a "OpaqueValueOfStruct" storage type thing, that perhaps we should just 'trust' that the user is doing something sensible with it.

I read that as you being in favor of trusting user to pass whatever type they want to `preferred_type`. I feel like I'm getting mixed signals from you on this topic, so it'd be nice if you can make yourself clear.

> I think the wording of the diagnostic is perhaps awkward, so we should bikeshed that as well

Sure.

> though I'm on the fence about the diagnostic existing at all. On one hand, this would be such a special case, as typically 'too large' for the type isn't an issue for any reason. On the other, the 'bool' case is perhaps uniquely an issue.

As we are going to attach semantic type information to our bit-fields, I see an opportunity to raise valid questions like "Why do you need 2 bits to hold a boolean value?". If intention is to reserve bits for future extensions, I believe it should be stated more explicitly via something along the lines of `enum E { False, True, Reserved = 3}`.

> What about if the person prefers the type be 'char', but has it be a size of 9? THIS is a case where it is reasonable (consider cases where you need to store the result of one of those calls to the C library that returns 'an int that is either a char or -1').

I'd like to state upfront that I'm diagnosing only booleans, because I see how it can help maintaining Clang headers. Now, on `CHAR_BIT == 8` platform, your example results in 8 bits for value, and 1 bit of padding (http://eel.is/c++draft/class.bit#1.sentence-6). I find such bit-field somewhat strange (why can't padding be declared explicitly via unnamed bit-field?), but maybe I haven't seen enough bit-fields in my life to appreciate it.

> So I guess I'm really questioning how accurate/error-free we could make that diagnostic.

In it's current form (only `bool` is checked), I consider it rather accurate. I'm yet to see a use case where `[[clang::preferred_type(bool)]] unsigned Predicate : 2` is the best solution available.

AaronBallman commented:
> > While I think that warning is accurate, I somewhat question the value of the 'bool' as working on this type
> 
> I'm not sure what you mean by "working" here, but I'd like to highlight that we have hundreds of single-bit bit-fields across Clang that would benefit from `[[clang::preferred_type(bool)]]`.

I wonder if we should treat one-bit bit-fields as if they were bool automatically (e.g., create this attribute implicitly in that case). How often do we expect to see one-bit bit-fields that are arithmetic? I'm sure it happens (to multiply against -1, 0, or 1 depending on the sign of the bit-field, for example), but I expect it to be quite rare compared to bool-like use.

(We could advance this patch as-is where it's explicit and add it implicitly in a follow-up.)

> > as, I'm not sure what it really means to put a non-enum here?
> 
> When it was `clang::debug_info_type`, the meaning was as simple as "I want the value of this bit-field to be interpreted as T by debuggers". Now that we moved to `preferred_type`, there might be more interpretations.
> 
> > What types ARE we allowing in this now?
> 
> As I mentioned in [#69104 (comment)](https://github.com/llvm/llvm-project/pull/69104#discussion_r1365269451), I'm not putting any restrictions on type parameter of the attribute, which makes even more sense for more generic `preferred_type`.

Hmmm, so I was thinking that the initial version of this attribute would be restricted to just integer types because we're only allowing the attribute to be applied to a bit-field and those require integer types. However, I think arbitrary types does make some sense (unless we want to get very specific). Consider: https://godbolt.org/z/esGhKoeW1

However, will this actually work in practice in the debugger? If not, perhaps we should limit to just integer and enumeration types for now, leaving the extension for the future.

erichkeane commented:
>As I mentioned in https://github.com/llvm/llvm-project/pull/69104#discussion_r1365269451, I'm not putting any restrictions on type parameter of the attribute, which makes even more sense for more generic preferred_type.

>But I'm confused by the fact you are raising this question. In https://github.com/llvm/llvm-project/pull/69104#discussion_r1364432624, you wrote:

It was just a question, based on the motivation I was misguided and wanted to make sure I had the complete list.  I can see value to leaving it unrestricted.

>I'd like to state upfront that I'm diagnosing only booleans, because I see how it can help maintaining Clang headers. Now, on CHAR_BIT == 8 platform, your example results in 8 bits for value, and 1 bit of padding (http://eel.is/c++draft/class.bit#1.sentence-6). I find such bit-field somewhat strange (why can't padding be declared explicitly via unnamed bit-field?), but maybe I haven't seen enough bit-fields in my life to appreciate it.

No, I'm saying that there ARE cases where the user would want to do something like:
` unsigned SomeChar : 9;`

That is because in some interfaces in the C standard library will return an integer that is EITHER a CHAR or an error-code.  So if the user is intending to capture that, this would be a false positive.  Definitively motivated IMO more than a 2 bit `bool`.  But since you're limiting this to `bool`, I'm less concerned.

>I wonder if we should treat one-bit bit-fields as if they were bool automatically (e.g., create this attribute implicitly in that case). How often do we expect to see one-bit bit-fields that are arithmetic? I'm sure it happens (to multiply against -1, 0, or 1 depending on the sign of the bit-field, for example), but I expect it to be quite rare compared to bool-like use.

I'd be tentative to do that initially, I think I'd prefer that be a future 'enhancement' after we evaluate the fall-out of this.  The rest of this patch has some significant promise, and I'd like to not hold it up trying to evaluate that.


>However, will this actually work in practice in the debugger? If not, perhaps we should limit to just integer and enumeration types for now, leaving the extension for the future.

OOof, yeah, I was initially thinking about something like `Fancy`, but the tests I'd need to feel comfortable doing something like that would be pretty extensive.  Additionally, anything 'floating point' would be REALLY messy as a bitfield anyway.  I like the idea of limiting this to `isIntegralOrEnumerationType` (or, perhaps that and `isa<EnumType>` so that we support incomplete enum types, just with no diagnostics).

Endilll committed:
Ensure that type argument is emitted in debug info

Endilll commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5910,6 +5910,51 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
+
+  TypeSourceInfo *ParmTSI = nullptr;
+  QualType QT = S.GetTypeFromParser(AL.getTypeArg(), &ParmTSI);
+  QT->getAsTagDecl()->setCompleteDefinitionRequired();
```
@dwblaikie Would it be fine to do this? The intention here is that we parse type argument of the attribute, and then use it in `CGDebugInfo.h` (you might be interested in changes there as well), bypassing Sema.

AaronBallman replied:
I think we want a different approach here. If we decide to do non-integral/enum types right now, I think we want to call `Sema::RequireCompleteType()` on the given type. This will then diagnose uses of types like `void` or `int[]` as well as forward declarations. Losing forward declarations is a bit sad (because it could be completed later), but I think we can relax this restriction in the future as we find uses.

Endilll replied:
I've changed this so that `preferred_type` accepts only complete types. We can lift this up later if it's too restrictive. I just don't want to deal with consequences of incomplete types in this PR.

Endilll commented:
> However, will this actually work in practice in the debugger? If not, perhaps we should limit to just integer and enumeration types for now, leaving the extension for the future.

I composed an example of that:
```cpp
struct A {
  short a1;
  short a2;
};

struct B {
  [[clang::preferred_type(A)]] unsigned b1 : 32 = 0x000F'000C;
};

int main()
{
    B b;
    return b.b1;
}
```
Nightly build of LLDB from apt.llvm.org handles it just fine:
```
Process 2755547 stopped
* thread #1, name = 'test-preferred-', stop reason = step in
    frame #0: 0x0000555555555148 test-preferred-type`main at test.cxx:13:14
   10   int main()
   11   {
   12       B b;
-> 13       return b.b1;
   14   }
(lldb) v -T
(B) b = {
  (A:32) b1 = {
    (short) a1 = 12
    (short) a2 = 15
  }
}
```
@erichkeane Do you still insist on restricting type argument to integral and enumeration types?

dwblaikie commented:
> ```c++
> struct A {
>   short a1;
>   short a2;
> };
> 
> struct B {
>   [[clang::preferred_type(A)]] unsigned b1 : 32 = 0x000F'000C;
> };
> 
> int main()
> {
>     B b;
>     return b.b1;
> }
> ```

An example where the layout doesn't match the normal struct layout might be more interesting? (like the reality of this situation ^ seems indistinguishable from similar code that used the real type and no bitfield, I think? (as far as the DWARF,struct layouts, etc, are concerned)) - maybe a struct with a `short` and a `char` (so it should have a byte of tail padding) then putting it in a 24 bit bitfield (maybe with an 8 bit bitfield after that - to check if attempts to write to the first bitfielded struct don't clobber the following bits) - or, maybe even a struct that ends in a bitfield so it's got less-than-whole-byte padding at the end, and see how that overlaps with embedding that as a bitfield in another struct with a subsequent bitfield.

I would guess a DWARF consumer might have more struggles there - though I Don't think it means we shouldn't do it, just expect to have to fix debuggers for this probably-rather-novel DWARF.

AaronBallman commented:
> > I wonder if we should treat one-bit bit-fields as if they were bool automatically (e.g., create this attribute implicitly in that case). How often do we expect to see one-bit bit-fields that are arithmetic? I'm sure it happens (to multiply against -1, 0, or 1 depending on the sign of the bit-field, for example), but I expect it to be quite rare compared to bool-like use.
> 
> I'd be tentative to do that initially, I think I'd prefer that be a future 'enhancement' after we evaluate the fall-out of this. The rest of this patch has some significant promise, and I'd like to not hold it up trying to evaluate that.

I'm retracting my suggestion there. I think this should be a default behavior of the debugger, but it shouldn't be forced on users by the compiler. Users who don't expect that behavior then won't have to fight against implicit compiler behavior.

dwblaikie commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3153,6 +3153,12 @@ def err_invalid_branch_protection_spec : Error<
+  "underlying type %0 of enumeration %1 doesn't match bit-field type %2">,
+  InGroup<BitFieldType>;
+def warn_attribute_bool_bitfield_width : Warning<
+  "bit-field that holds a boolean value should have width of 1 instead of %0">,
```
I think any case where the bitfield is actually /larger/ than the representational range would be potentially useful warning. The user could use an anonymous bitfield if they want to pad something out instead between bitfield elements instead?

dwblaikie replied:
Questionable whether this has sufficient bug-finding power. I mean, they're probably bugs, and the bug is that your struct ends up larger than you intended... which,  yeah, if you're packing things, then that's perhaps a useful diagnostic.

Endilll replied:
> The user could use an anonymous bitfield if they want to pad something out instead between bitfield elements instead?

I share this point of view, but Erich mentioned why oversized bit-fields are useful: https://github.com/llvm/llvm-project/pull/69104#issuecomment-1772828263

AaronBallman replied:
Stepping back a bit, I think there are two orthogonal goals here: 1) warn the user "hey, this bit-field is too small to fit everything of that type you said you'll be storing in here" and 2) warn the user "hey, this bit-field is larger than it needs to be in order to fit everything of that type you said you'll be storing in here."

I think #1 is better-served by the existing warning that happens at a problematic expression rather than trying to do it at the point of declaration: https://godbolt.org/z/MG8sPEq6o  This won't help in this situation: https://godbolt.org/z/8sdEY559K but that situation is one the user is explicitly asking to lose type information and so it seems reasonable that we don't diagnose (IMO).

I think #2 is really more of a clang-tidy concern because it's not so much about correctness but space efficiency. We've got things like `-Wpadded` to help point out those issues and maybe this attribute could feed into a similar diagnostic in Clang, but I'm not convinced it will find enough issues in practice to warrant adding it.

So I currently lean towards not diagnosing this situation. (I also wonder if we're accidentally scope-creeping this PR somewhat -- what kicked all this off was me suggesting to rename the attribute so it could be extensible to situations like this, but I didn't intend for that to mean "right now" FWIW.)

Endilll replied:
Makes sense.
I have removed `bool` diagnostic.

Endilll committed:
Require type argument to be a complete type

erichkeane commented:
> > However, will this actually work in practice in the debugger? If not, perhaps we should limit to just integer and enumeration types for now, leaving the extension for the future.
> 
> I composed an example of that:
> 
> ```c++
> struct A {
>   short a1;
>   short a2;
> };
> 
> struct B {
>   [[clang::preferred_type(A)]] unsigned b1 : 32 = 0x000F'000C;
> };
> 
> int main()
> {
>     B b;
>     return b.b1;
> }
> ```
> 
> Nightly build of LLDB from apt.llvm.org handles it just fine:
> 
> ```
> Process 2755547 stopped
> * thread #1, name = 'test-preferred-', stop reason = step in
>     frame #0: 0x0000555555555148 test-preferred-type`main at test.cxx:13:14
>    10   int main()
>    11   {
>    12       B b;
> -> 13       return b.b1;
>    14   }
> (lldb) v -T
> (B) b = {
>   (A:32) b1 = {
>     (short) a1 = 12
>     (short) a2 = 15
>   }
> }
> ```
> 
> @erichkeane Would you still prefer restricting type argument to integral and enumeration types?

I think that is the least dangerous, but the `requireCompleteType` isn't awful (and I can do a disagree & commit on).  I think it gets goofy when the sizes don't match with floating point/complex/etc, but I also see "you got what you asked for!" as being a reasonable defense to that.

Endilll committed:
Remove diagnostic for bool bif-fields

Endilll committed:
Fix formatting

Endilll commented:
> but I also see "you got what you asked for!" as being a reasonable defense to that.

That's my thinking indeed, and the reason why I opposed to Aaron's proposal to implicitly mark 1-bit bit-fields as `preferred_type(bool)`.

AaronBallman commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7219,6 +7219,31 @@ its underlying representation to be a WebAssembly ``funcref``.
+Without the attribute, a debugger is likely to display the value `1` for `ColorVal`
+and `0` for `UseAlternateColorSpace`. With the attribute, the debugger may now
+display `Green` and `false` instead.
+  }];
```
We should add additional documentation now that we've not limited this to just enumeration types. Maybe something along the lines of:
```

This can be used to map a bit-field to an arbitrary type that isn't integral or an enumeration type. For example:

.. code-block:: c++

    struct A {
      short a1;
      short a2;
    };

    struct B {
      [[clang::preferred_type(A)]] unsigned b1 : 32 = 0x000F'000C;
    };

will associate the type ``A`` with the ``b1`` bit-field and is intended to display something like this in the debugger:

.. code-block:: text

    Process 2755547 stopped
    * thread #1, name = 'test-preferred-', stop reason = step in
        frame #0: 0x0000555555555148 test-preferred-type`main at test.cxx:13:14
       10   int main()
       11   {
       12       B b;
    -> 13       return b.b1;
       14   }
    (lldb) v -T
    (B) b = {
      (A:32) b1 = {
        (short) a1 = 12
        (short) a2 = 15
      }
    }

Note that debuggers may not be able to handle more complex mappings and so this usage is debugger-dependent.
```

Endilll replied:
Done!

AaronBallman commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7219,6 +7219,31 @@ its underlying representation to be a WebAssembly ``funcref``.
+      [[clang::preferred_type(bool)]] unsigned UseAlternateColorSpace : 1;
+    } s = { Green, false };
+
+Without the attribute, a debugger is likely to display the value `1` for `ColorVal`
+and `0` for `UseAlternateColorSpace`. With the attribute, the debugger may now
+display `Green` and `false` instead.
```
```suggestion
Without the attribute, a debugger is likely to display the value ``1`` for ``ColorVal``
and ``0`` for ``UseAlternateColorSpace``. With the attribute, the debugger may now
display ``Green`` and ``false`` instead.
```

Endilll replied:
Applied.

AaronBallman reviewed (commented):
The changes should also come with a release note, but I'm generally happy with the code. I did have some suggestions for changes to documentation though.

AaronBallman commented:
> The changes should also come with a release note, but I'm generally happy with the code. I did have some suggestions for changes to documentation though.

Suggestion for the release note would be something along these lines:
```
- Clang now supports ``[[clang::preferred_type(type-name)]]`` as an attribute
which can be applied to a bit-field. This attribute helps to map a bit-field
back to a particular type that may be better-suited to representing the bit-
field but cannot be used for other reasons and will impact the debug
information generated for the bit-field. This is most useful when mapping a
bit-field of basic integer type back to a ``bool`` or an enumeration type.
e.g.,

.. code-block:: c++

    enum E { Apple, Orange, Pear };
    struct S {
      [[clang::preferred_type(E)]] unsigned FruitKind : 2;
    };

When viewing ``S::FruitKind`` in a debugger, it will behave as if the member
was declared as type ``E`` rather than ``unsigned``.
```

Endilll committed:
Add a release note

Endilll committed:
Update the documentation using Aaron's suggestions

Endilll commented:
Thank you @AaronBallman for writing even more documentation!

AaronBallman reviewed (approved):
These changes LGTM, but please give @erichkeane a chance to chime in given his previous feedback.

erichkeane reviewed (approved):
LGTM, i think we're in an acceptable way forward, particularly on diagnostics.
